### PR TITLE
[PT-D][Sequence Parallelism] Enable DTensor based Native sequence parallelism

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94369

### CODE REVIEW

File: torch/distributed/tensor/parallel/style.py [hunk 1]
```diff
@@ -55,6 +56,21 @@ def __init__(self) -> None:
...
+    pair like what Megatron-LM(https://arxiv.org/abs/1909.08053) is doing.
+    We assume both input and output needs to a replicate DTensor.
+
+    .. warning::
+        PairwiseParallel only supports ``nn.Multihead Attention``,
+        ``nn.Transformer`` or even-number-layer MLP for now.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(make_input_replicate_1d, make_output_shard_1d)
```
Comment (fduwjj) [inline torch/distributed/tensor/parallel/style.py:71]: I am ok with moving converting tensor to Shard DTensor and to_local into prepare_input and prepare_out. Open for suggestion.

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -55,6 +56,21 @@ def __init__(self) -> None:
         super().__init__(make_input_replicate_1d, make_output_tensor)
 
 
+class PairwiseSequenceParallel(ParallelStyle):
+    """
+    PairwiseParallel concatenate colwise and rowwise styles as a fixed
+    pair like what Megatron-LM(https://arxiv.org/abs/1909.08053) is doing.
+    We assume both input and output needs to a replicate DTensor.
```
Comment (mrshenli) [inline torch/distributed/tensor/parallel/style.py:63]: both input and output are replicate DTensors? or both input and output needs to become a replicate DTensor?

File: torch/distributed/tensor/parallel/style.py (hunk 1 above)
Comment (mrshenli) [reply torch/distributed/tensor/parallel/style.py:71]: Does this align with the comments above? This is converting input to replicate and output to shard, but docstring mentioned both of them needs to be replicate DTensor?

File: torch/distributed/tensor/parallel/api.py [hunk 3]
```diff
@@ -377,9 +378,11 @@ def _parallelize_mlp(
     .. warning::
         We only support ``PairwiseParallel`` right now.
```
Comment (mrshenli) [inline torch/distributed/tensor/parallel/api.py:379]: update this accordingly?

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 4]
```diff
@@ -140,6 +142,115 @@ def test_mlp_megatron_e2e(self):
...
+            ).to_local(),
+        )
+        self.assertEqual(
+            model.net2.weight,
+            model_tp.net2.weight.redistribute(
+                device_mesh=device_mesh, placements=replicate
+            ).to_local(),
+        )
+        # Due to the trick we use for Partial aggregation, we only check the weight when local_rank = 0.
+        if self.rank == 0:
```
Comment (wanchaol) [inline test/distributed/tensor/parallel/test_tp_examples.py:235]: curious why we need to test this as I thought it's already been tested above?

**Review by wanchaol — COMMENTED** [review]
Nice work! Looks like we should also make input be shard on batch iiuc? See comments inlined.

File: torch/distributed/tensor/parallel/style.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:71]: Nope, I tried to make input Sharded, but during sharding propagation, we don't do redistribute for input DTensor. I run into errors when doing this and this is the finding I learned from debugging.

File: torch/distributed/_tensor/redistribute.py [hunk 5]
```diff
@@ -226,6 +226,9 @@ def backward(ctx, grad_output: "dtensor.DTensor"):  # type: ignore[override]
             if current.is_replicate() and target.is_partial():
                 # keep target placement to replicate instead of partial in this case
                 target_placements.append(current)
+            elif current.is_shard() and target.is_partial():
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_tensor/redistribute.py:229]: Sure.

File: test/distributed/tensor/parallel/test_tp_examples.py (hunk 4 above)
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_tp_examples.py:235]: This the bias for net2. Above we only tested its grad

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 6]
```diff
@@ -140,6 +142,115 @@ def test_mlp_megatron_e2e(self):
...
+            self.assertEqual(
+                model.net2.bias,
+                model_tp.net2.bias.redistribute(
+                    device_mesh=device_mesh, placements=replicate
+                ).to_local(),
+            )
+
+        inp = torch.rand(*inp_size, device=self.device_type)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_tp_examples.py:243]: This is to test the FWD again after optimizer update.

File: torch/distributed/tensor/parallel/style.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:63]: Opps.. I forgot to update the comments. To answer your question, both input and output should be a Sharded DTensor, but in reality we don't do redistribute for input DTensor (From Sharded to Replicated) in matmul.. Let me think how we can better approach this.

File: torch/distributed/tensor/parallel/style.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:71]: Thanks for calling that out. This is still in prototyping, and we need to make a decision here. Let me update the PR.

File: torch/distributed/tensor/parallel/style.py [hunk 7]
```diff
@@ -44,15 +45,36 @@ class PairwiseParallel(ParallelStyle):
     """
     PairwiseParallel concatenate colwise and rowwise styles as a fixed
     pair like what Megatron-LM(https://arxiv.org/abs/1909.08053) is doing.
-    We assume both input and output needs to a replicate DTensor.
+    We assume both input and output need to a replicate DTensor.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:48]:
```suggestion
    We assume both input and output need to be replicate DTensors.
```

File: torch/distributed/tensor/parallel/style.py [hunk 8]
```diff
@@ -121,7 +143,7 @@ def make_input_shard_1d_dim_last(
 
     Args:
         input (Union[:class:`torch.Tensor`, :class:`DTensor`]):
-            This single tensor will be sharded on dimension ``dim``
+            This single tensor will be sharded on dimension ``-1``
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:146]: nit: will be sharded on the last dimension

File: torch/distributed/tensor/parallel/style.py [hunk 9]
```diff
@@ -131,11 +153,38 @@ def make_input_shard_1d_dim_last(
...
     """
     return make_input_shard_1d(input, device_mesh, dim=-1)  # type: ignore[call-arg]
 
 
+def make_input_reshard_replicate(
+    input: torch.Tensor,
+    device_mesh: DeviceMesh,
+) -> DTensor:
+    """
+    To convert a Tensor on different ranks to a Sharded DTensor
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:166]:
```suggestion
    To construct a Sharded DTensor from a Tensor on different ranks
```

File: torch/distributed/tensor/parallel/style.py [hunk 10]
```diff
@@ -131,11 +153,38 @@ def make_input_shard_1d_dim_last(
...
+    and then convert to a replicate DTensor.
+
+    Args:
+        input (:class:`torch.Tensor`):
+            This single tensor consists of a global DTensor sharded on dimension ``0``
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:171]: why a single tensor would consists a global DTensor?

I guess it should be sth like: "The input tensor on each rank which could form a global DTensor ..."

File: torch/distributed/tensor/parallel/style.py [hunk 11]
```diff
@@ -44,15 +45,36 @@ class PairwiseParallel(ParallelStyle):
...
+    We assume both input and output need to a sharded DTensor.
+
+    .. warning::
+        PairwiseSequenceParallel only supports ``nn.Multihead Attention``,
+        ``nn.Transformer`` or even-number-layer MLP for now.
+    """
+
     def __init__(self) -> None:
-        super().__init__(make_input_replicate_1d, make_output_tensor)
+        super().__init__(make_input_reshard_replicate, make_output_shard_1d)
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:77]: For output, we are transforming it to Shard(0) DTensor, shouldn't we also transform it to a torch.Tensor like the PiarwiseParallel?

Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:77]: OK, we then need a new output function. Let me add it.

**Comment by fduwjj** [comment]
manually rebase and resolved the conflict.

**Comment by fduwjj** [comment]
Fix the error in the test and the rest of CI failure are not related to this PR.

File: torch/distributed/tensor/parallel/style.py [hunk 12]
```diff
@@ -240,3 +292,27 @@ def make_output_tensor(
...
     ).to_local()  # type: ignore[call-arg]
+
+
+@_prepare_output_validate  # type: ignore[arg-type] # pyre-ignore[56]
+def make_output_reshard_tensor(
+    output: DTensor,
+    device_mesh: Optional[DeviceMesh] = None,
+) -> torch.Tensor:
+    """
+    Convert Output DTensor to a sharded DTensor and return the local tensor.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:303]: nit: we should mention it's a "sharded on batch dimension DTensor"

File: torch/distributed/tensor/parallel/style.py [hunk 13]
```diff
@@ -131,11 +155,39 @@ def make_input_shard_1d_last_dim(
...
     return make_input_shard_1d(input, device_mesh, dim=-1)  # type: ignore[call-arg]
 
 
+@_prepare_input_validate  # type: ignore[arg-type] # pyre-ignore[56]
+def make_input_reshard_replicate(
+    input: torch.Tensor,
+    device_mesh: DeviceMesh,
+) -> DTensor:
+    """
+    To construct a Sharded DTensor from a tensor on different ranks
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:169]: nit: we should mention it's a "sharded on batch dimension DTensor"

**Review by wanchaol — APPROVED** [review]
Great work! Only have some nits

**Pull request closed without merging**